Firestore/Source/Local/FSTMemoryPersistence.mm

```diff
@@ -105,4 +143,226 @@ - (void)shutdown {
+                                                          sequenceNumber:_currentSequenceNumber];
+  [_persistence.queryCache updateQueryData:updated];
+}
+
+- (void)limboDocumentUpdated:(const DocumentKey &)key {
+  _sequenceNumbers[key] = self.currentSequenceNumber;
+}
+
+- (void)startTransaction:(absl::string_view)label {
+  _currentSequenceNumber = [_listenSequence next];
+}
+
+- (void)commitTransaction {
+  _currentSequenceNumber = kFSTListenSequenceNumberInvalid;
+}
+
+- (void)enumerateTargetsUsingBlock:(void (^)(FSTQueryData *queryData, BOOL *stop))block {
+  return [_persistence.queryCache enumerateTargetsUsingBlock:block];
+}
+
+- (void)enumerateMutationsUsingBlock:
+    (void (^)(const DocumentKey &key, FSTListenSequenceNumber sequenceNumber, BOOL *stop))block {
+  BOOL stop = NO;
+  for (auto it = _sequenceNumbers.begin(); !stop && it != _sequenceNumbers.end(); ++it) {
+    FSTListenSequenceNumber sequenceNumber = it->second;
+    const DocumentKey &key = it->first;
+    if (![_persistence.queryCache containsKey:key]) {
+      block(key, sequenceNumber, &stop);
+    }
+  }
+}
+
+- (int)removeTargetsThroughSequenceNumber:(FSTListenSequenceNumber)sequenceNumber
+                              liveQueries:(NSDictionary<NSNumber *, FSTQueryData *> *)liveQueries {
+  return [_persistence.queryCache removeQueriesThroughSequenceNumber:sequenceNumber
+                                                         liveQueries:liveQueries];
+}
+
+- (int)removeOrphanedDocumentsThroughSequenceNumber:(FSTListenSequenceNumber)upperBound {
+  return [(FSTMemoryRemoteDocumentCache *)_persistence.remoteDocumentCache
+      removeOrphanedDocuments:self
+        throughSequenceNumber:upperBound];
+}
+
+- (void)addReference:(const DocumentKey &)key {
+  _sequenceNumbers[key] = self.currentSequenceNumber;
+}
+
+- (void)removeReference:(const DocumentKey &)key {
+  _sequenceNumbers[key] = self.currentSequenceNumber;
+}
+
+- (BOOL)mutationQueuesContainKey:(const DocumentKey &)key {
+  const MutationQueues &queues = [_persistence mutationQueues];
+  for (auto it = queues.begin(); it != queues.end(); ++it) {
+    if ([it->second containsKey:key]) {
+      return YES;
+    }
+  }
+  return NO;
+}
+
+- (void)removeMutationReference:(const DocumentKey &)key {
+  _sequenceNumbers[key] = self.currentSequenceNumber;
+}
+
+- (BOOL)isPinnedAtSequenceNumber:(FSTListenSequenceNumber)upperBound
+                        document:(const DocumentKey &)key {
+  if ([self mutationQueuesContainKey:key]) {
+    return YES;
+  }
+  if ([_additionalReferences containsKey:key]) {
+    return YES;
+  }
+  if ([_persistence.queryCache containsKey:key]) {
+    return YES;
+  }
+  auto it = _sequenceNumbers.find(key);
+  if (it != _sequenceNumbers.end() && it->second > upperBound) {
+    return YES;
+  }
+  return NO;
+}
+
+@end
+
+@implementation FSTMemoryEagerReferenceDelegate {
+  std::unique_ptr<std::unordered_set<DocumentKey, DocumentKeyHash> > _orphaned;
```

Comment:
With C++11, the lexical structure changed such that `>>` to close off nested templates is allowed. You no longer need to protect against this here.

Reply:
We are living in the future!